.changeset/witty-ears-repair.md

```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/backend-test-utils': minor
+---
+
+**BREAKING** Removed support for PostgreSQL versions 9 and 11, also added support for versions 15 and 16
```

Comment:
As mentioned elsewhere, posting here for posterity:

We probably should keep the old IDs around (possibly indefinitely, only ever adding to the list), to not break existing users. However, the default construction method for `TestDatabases` should get the breaking change that its default behavior is to follow the pg policy of the oldest and newest LTS version + sqlite. This makes it trivial for users to follow our testing guidelines out of the box, without having to specify and then update `ids`. Then there can be an `extraIds` parameter for those extracurricular IDs like mysql on top of that, that you can add without losing out on the default-guideline behavior.

Test in this repo that currently supply `ids` in this repo, should then be updated to use that new `extraIds` instead; essentially no test anywhere in here should probably use `ids`.

Reply:
We talked a bit more about this this morning. Realising that for an adopter, we have to consider the step of moving to a new database major a large and potentially costly undertaking. Potentially they could need to requisition a new database server to test against etc, depending on how they do things. So we can probably never make the `create` method "silently" change _any_ behaviour when you upgrade Backstage - at least in CI.

The existing old behaviour of `TestDatabases` is probably already good enough for the CI use case, in that it automatically only runs for things that you DID opt in to with the env var. That lets adopters gradually migrate to whatever set of supported databases they need, at their own pace, even when calling the `create()` method entirely without parameters.

The trickier case then is, what to do when running tests locally and/or in "docker mode". Currently it runs ALL possible databases when `CI=1` is set, which has benefits (coverage) and drawbacks (performance, lowest-common-denominator feature support, testing differently in CI and local, etc). And if you supply a list of ids, then you start having to carefully match that list of IDs with what your CI environment actually has env vars for. 🤔

We discussed whether perhaps you could have a static method that you can call in `setupTests.ts`, like `TestDatabases.setDefaults({ ids: TestDatabaseId[] })`. That way you could at least just do it once per package to declare what coverage you want for all tests in there. Updating that when moving database versions, is less work than doing it in all individual test files. Not groundbreaking, but a convenience factor nonetheless.

Reply:
Alright, got back to this today. I've added back the old Ids and got the tests passing.

I'd like to no tackle the idea of adding "a static method that you can call in `setupTests.ts`, like `TestDatabases.setDefaults({ ids: TestDatabaseId[] })`" but I'm not totaly sure how that would look so if you could give me some more details that would help. Going to run with this for a bit today and follow up if I get anywhere 👍

Reply:
We can do that separately! I'll need to think a bit on it.

Reply:
Ah, ok, I missed this message. I just pushed a change for that, it does exactly as described. You add this to your `setupTests.ts`:

```ts
TestDatabases.setDefaults({
  ids: ['SQLITE_3', 'POSTGRES_12', 'POSTGRES_16'],
});
```

and then if your `TestDatabases.create()` doesn't pass in any `ids` it will use the default above, if you do have `ids` it will create a unique list of the defaults and the provided `ids`

I do think my code could be improved/refactored I got hung up around the `Object.assign()` call

What do you think @freben?

Reply:
Resolving this as there are follow up comments related to it 👍